### PR TITLE
Multi-cursor components: queue iterator

### PR DESCRIPTION
**Pull request by yycptt**
**What changed?**
- Add history queue scope definition and related methods

**Why?**
- Building blocks for multi-cursor queue

**How did you test it?**
- Unit test

**Potential risks**
- N/A, not used

**Is hotfix candidate?**
- No

### CODE REVIEW

File: service/history/queues/iterator.go [hunk 1]
```diff
@@ -0,0 +1,145 @@
...
+		paginationFnProvider paginationFnProvider
+		remainingRange       tasks.Range
+
+		pagingIterator collection.Iterator[tasks.Task]
+	}
+)
+
+func NewIterator(
+	paginationFnProvider paginationFnProvider,
+	r tasks.Range,
```
Comment (wxing1292) [inline service/history/queues/iterator.go:58]: `r tasks.Range,` -> `range tasks.Range,`

File: service/history/queues/iterator.go [hunk 2]
```diff
@@ -0,0 +1,145 @@
...
+	if err != nil {
+		return nil, err
+	}
+
+	taskKey := task.GetKey()
+	taskID := taskKey.TaskID
+	if taskKey.FireTime.UnixNano() <= 0 {
+		// only increase taskID for immediate task category
+		taskID++
+	}
```
Comment (wxing1292) [inline service/history/queues/iterator.go:92]: why not define Next() on task key and call next here?

why only handle transfer task IDs? why timer task ID is not "++"ed?

Comment (yycptt) [reply service/history/queues/iterator.go:92]: Synced offline, will always increase taskID and overflow to fire time if taskID is maxInt64. 

When querying db though, taskID used should be 0 for scheduled tasks (due to sanity check in persistence layer). This logic will be inside the function returned by `paginationFnProvider` so technically unrelated to iterator itself.

File: service/history/queues/iterator.go (hunk 1 above)
Comment (yycptt) [reply service/history/queues/iterator.go:58]: `range` is reserved keyword.

**Review by wxing1292 — APPROVED** [review]

File: service/history/tasks/key.go [hunk 3]
```diff
@@ -86,6 +87,26 @@ func (left Key) CompareTo(right Key) int {
...
+	}
+	return NewKey(k.FireTime, k.TaskID-1)
+}
+
+func (k Key) Next() Key {
+	if k.TaskID == math.MaxInt64 {
+		if k.FireTime.UnixNano() == math.MaxInt64 {
+			panic("Key encountered positive overflow")
+		}
+		return NewKey(k.FireTime.Add(time.Nanosecond), 0)
```
Comment (yiminc) [inline service/history/tasks/key.go:105]: do you plan to iterate by 1ns at a time for timer queue range?

Comment (yycptt) [reply service/history/tasks/key.go:105]: no, we will never reach here. If so, it means the taskID for this shard is already max int64 and the shard can't be used anymore. 

So yes we technically don't need this logic. But Wenquan suggests that this make more sense logically if you consider fire time as the high 64 bits of the task key. And when the lower 64bits (taskID) overflows, we add one to the fire time.

**Pull request merged**